crates/ra_hir/src/adt.rs

```diff
@@ -0,0 +1,213 @@
+use std::sync::Arc;
+
+use ra_syntax::{SmolStr, ast::{self, NameOwner, StructFlavor}};
+
+use crate::{
+    DefId, Cancelable,
+    db::{HirDatabase},
+    module::Module,
+    ty::{Ty},
+};
+
+pub struct Struct {
+    def_id: DefId,
+}
+
+impl Struct {
+    pub(crate) fn new(def_id: DefId) -> Self {
+        Struct { def_id }
+    }
+
+    pub fn def_id(&self) -> DefId {
+        self.def_id
+    }
+
+    pub fn variant_data(&self, db: &impl HirDatabase) -> Cancelable<Arc<VariantData>> {
+        Ok(db.struct_data(self.def_id)?.variant_data.clone())
+    }
+
+    pub fn struct_data(&self, db: &impl HirDatabase) -> Cancelable<Arc<StructData>> {
+        Ok(db.struct_data(self.def_id)?)
+    }
+
+    pub fn name(&self, db: &impl HirDatabase) -> Cancelable<SmolStr> {
+        Ok(db.struct_data(self.def_id)?.name.clone())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct StructData {
+    name: SmolStr,
+    variant_data: Arc<VariantData>,
+}
+
+impl StructData {
+    pub(crate) fn new(
+        db: &impl HirDatabase,
+        module: &Module,
+        struct_def: ast::StructDef,
+    ) -> Cancelable<StructData> {
+        let name = struct_def
+            .name()
+            .map(|n| n.text())
+            .unwrap_or(SmolStr::new("[error]"));
```

Comment:
Interesting! It seems easier to just not have anonymous items in the hir. But then you'll be unable to get the type of the field within the struct itself. So perhaps having a sort-of NaN name would work better?

Reply:
OTOH, we don't need  "to get the type of the field within the struct itself" so let's perhaps just return `None` if the name is missing?

Reply:
Hm. I think I'll make the name an option. We might still want to get the HIR for a struct without name (though it's admittedly unlikely).